### PR TITLE
Prevent timer drift

### PR DESCRIPTION
**Pull request by ccoors**
## Prevent timer drift

Type A:
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor/Code Cleanup
- [ ] Docs
- [ ] Capability implementation for existing core capability
- [ ] New robot implementation

# Description (Type A)

As recently [discussed in the Valetudo channel](https://t.me/c/1357472831/52008), `setInterval` definitely drifts over time, which could cause timers to be missed. This PR should fix the issue.

It does so by comparing the current time to the time it last ran and evaluating the timers an appropriate number of times. Since absolute precision is not needed anyway, this solution is less wasteful of CPU cycles than the idea of using the driftless package, discussed [here](https://t.me/c/1357472831/51983).

### CODE REVIEW

File: backend/lib/scheduler/Scheduler.js [hunk 1]
```diff
@@ -18,22 +20,16 @@ class Scheduler {
         this.config = options.config;
         this.robot = options.robot;
         this.ntpClient = options.ntpClient;
+        this.checkedTime = null;
```
Comment (Hypfer) [inline backend/lib/scheduler/Scheduler.js:23]: I needed a while to understand what is happening here.
This naming change might help, unless I've misunderstood how this works

```suggestion
        this.nextScheduledCheckTime = null;
```

File: backend/lib/scheduler/Scheduler.js [hunk 2]
```diff
@@ -48,17 +44,40 @@ class Scheduler {
...
-            ) {
-                Logger.info("Executing timer " + timerDefinition.id);
-                this.executeTimer(timerDefinition);
-            }
-        });
+        const nowTime = new Date().getTime();
+
+        // checkedTime is invalid, either because we just started or because NTP moved
+        // us too far in time. In either case, make the following loop run only once.
+        if (this.checkedTime === null || this.checkedTime > nowTime || this.checkedTime < (nowTime - 60 * MS_PER_MIN)) {
```
Comment (Hypfer) [inline backend/lib/scheduler/Scheduler.js:51]:
```suggestion
        if (this.nextScheduledCheckTime === null || this.nextScheduledCheckTime > nowTime || this.nextScheduledCheckTime < (nowTime - 60 * MS_PER_MIN)) {
```

File: backend/lib/scheduler/Scheduler.js [hunk 3]
```diff
@@ -48,17 +44,40 @@ class Scheduler {
...
+            this.checkedTime = nowTime - MS_PER_MIN;
```
Comment (Hypfer) [inline backend/lib/scheduler/Scheduler.js:52]:
```suggestion
            this.nextScheduledCheckTime = nowTime - MS_PER_MIN;
```

File: backend/lib/scheduler/Scheduler.js [hunk 4]
```diff
@@ -48,17 +44,40 @@ class Scheduler {
...
+        }
+
+        // setInterval drifts. It has to, because that's how the JS event loop works.
+        // That could cause timer events to be missed, e.g. when evaluateTimers() is
+        // called in minute 10 and 59 seconds and the next time in minute 12 and 00
+        // seconds. To counteract this race condition, the following loop runs twice
+        // in such a case.
+        while (this.checkedTime < nowTime) {
```
Comment (Hypfer) [inline backend/lib/scheduler/Scheduler.js:60]:
```suggestion
        while (this.nextScheduledCheckTime < nowTime) {
```

File: backend/lib/scheduler/Scheduler.js [hunk 5]
```diff
@@ -48,17 +44,40 @@ class Scheduler {
...
+            this.checkedTime += MS_PER_MIN;
```
Comment (Hypfer) [inline backend/lib/scheduler/Scheduler.js:61]:
```suggestion
            this.nextScheduledCheckTime += MS_PER_MIN;
```

**Comment by ccoors** [comment]
Renamed it, but that required `this.nextScheduledCheckTime` to be updated at the end of the loop instead. Otherwise the naming is wrong.

**Pull request merged**